routing/matcher.go

```diff
@@ -268,101 +267,28 @@ func addTreeMatchers(pathTree *pathmux.Tree, matchers map[string]*pathMatcher, s
 func addLeafToPath(pms map[string]*pathMatcher, path string, l *leafMatcher) {
 	pm, ok := pms[path]
 	if !ok {
-		pm = &pathMatcher{freeWildcardParam: freeWildcardParam(path)}
+		pm = &pathMatcher{}
 		pms[path] = pm
 	}
 
 	pm.leaves = append(pm.leaves, l)
 }
 
-func moveToSubtreeIfExists(subtree *pathMatcher, paths map[string]*pathMatcher, path string) {
-	pm, exists := paths[path]
-	if !exists {
-		return
-	}
-
-	subtree.leaves = append(subtree.leaves, pm.leaves...)
-	delete(paths, path)
-}
-
-// when a path subtree and a path with free wildcard has the same root,
-// we need to merge them to the same leaf, but we need to preserve the
-// original parameter name. The tree accepts only a single wildcard
-// param name for a given position, so we need to store the original
-// param name elsewhere (subtreeMergeControl).
-func mergeWithSubtreeIfConflicting(subtrees, paths map[string]*pathMatcher) {
-	subtreesUnnamed := make(map[string]*pathMatcher)
-	subtreesFullPaths := make(map[string]string)
-	for stp, stm := range subtrees {
-		// trim the param name, if exists
-		var subtreePath string
-		stpParam := freeWildcardRx.FindString(stp)
-		if stpParam == "" {
-			subtreePath = stp
-		} else {
-			subtreePath = stp[:len(stp)-len(stpParam)]
-		}
-
-		if subtreePath == "" {
-			subtreePath = "/"
-		}
-
-		subtreesUnnamed[subtreePath] = stm
-		subtreesFullPaths[subtreePath] = stp
-	}
-
-	var removeFromPaths []string
-	for p, pm := range paths {
-		fwm := freeWildcardRx.FindString(p)
-		if fwm == "" {
-			continue
-		}
-
-		subRoot := p[:len(p)-len(fwm)]
-		if subRoot == "" {
-			subRoot = "/"
-		}
-
-		stm, ok := subtreesUnnamed[subRoot]
-		if !ok {
-			continue
-		}
-
-		fromParam := stm.freeWildcardParam
-		if fromParam == "" {
-			fromParam = "*"
-		}
-
-		for _, l := range pm.leaves {
-			l.subtreeMergeControl.ignoreSubtreeRoot = true
-			l.subtreeMergeControl.subtreeRoot = subRoot
-			l.subtreeMergeControl.mapParamNameFrom = fromParam
-			l.subtreeMergeControl.mapParamNameTo = pm.freeWildcardParam
+func addSubtreeLeafsToPath(pms map[string]*pathMatcher, path string, l *leafMatcher, o MatchingOptions) {
+	basePath := freeWildcardRx.ReplaceAllLiteralString(path, "")
+	basePath = strings.TrimSuffix(basePath, "/")
```

Comment:
I am not sure if the function works like expected. Maybe yes, maybe not.

https://play.golang.org/p/_tTUtfhO5q1
```
	freeWildcardRx := regexp.MustCompile("/[*][^/]*$")
	s := freeWildcardRx.ReplaceAllLiteralString("/foo/*bar/*baz", "")
	fmt.Println(s) // "/foo/*bar"
```

Reply:
This is the intended behavior. Remove the free wildcard from the end in order to get a base path, without free wildcard params.
For a `PathSubtree`  we need to add matchers for the `basePath` and for `basePath + free wildcard`. You also add `basePath + "/"` depending on the trailing slash flag.

Reply:
In your example, `/foo/*bar/*baz`, `baz` is a free wildcard param, while `bar` is a normal wildcard.
`/foo/*bar/*baz` is equivalent to `/foo/:bar/*baz`

Reply:
ok thanks!